File: crane/crane.py
import itertools as it
import functools as ft
import operator as op
import collections as co
import fractions as fr
import math

vec = co.namedtuple("vec", "x y")

make_vec_op = lambda o: lambda self, other: vec(
    *it.starmap(o, zip(self, other))
)
make_scalar_op = lambda o: lambda self, other: vec(
    *map(ft.partial(o, other), self)
)

vec.__add__ = make_vec_op(op.add)
vec.__sub__ = make_vec_op(op.sub)
vec.__mul__ = make_scalar_op(op.mul)
vec.__truediv__ = make_scalar_op(lambda a, b: b / a)
vec.__rmul__ = vec.__mul__
cross = lambda a, b: a.x * b.y - a.y * b.x
area3 = lambda a, b, c: cross(b - a, c - a) / 2
centroid3 = lambda a, b, c: (a + b + c) / 3

area = lambda poly: sum(
    cross(poly[i], poly[(i + 1) % len(poly)]) / 2
    for i in range(len(poly))
)

area_relative = lambda poly: sum(
    area3(poly[0], poly[i], poly[i + 1])
    for i in range(1, len(poly) - 1)
)

centroid = lambda poly: ft.reduce(
    op.add,
    (
        cross(poly[i], poly[(i + 1) % len(poly)])
        / 2
        * (poly[i] + poly[(i + 1) % len(poly)])
        / 3
        for i in range(len(poly))
    ),
) / area(poly)


centroid_relative = lambda poly: ft.reduce(
    op.add,
    (
        area3(poly[0], poly[i], poly[i + 1])
        * centroid3(poly[0], poly[i], poly[i + 1])
        for i in range(1, len(poly) - 1)
    ),
) / area_relative(poly)


def min_max(l):
    l = list(l)
    return min(l), max(l)


solve_inequality = lambda a, b: (
    ((0, math.inf) if 0 <= b else (0, -1))
    if a == 0
    else (0, b / a)
    if a > 0
    else (max(b / a, 0), math.inf)
)


intersect = lambda i, j: (max(i[0], j[0]), min(i[1], j[1]))


def solve(poly):
    x = poly[0].x
    l, r = min_max(p.x for p in poly if p.y == 0)
    c = centroid(poly).x
    m = abs(area(poly))

    a, b = intersect(
        solve_inequality(x - r, m * (r - c)),
        solve_inequality(l - x, m * (c - l)),
    )

    return (
        (math.floor(a), math.inf if b == math.inf else math.ceil(b))
        if a <= b
        else None
    )


def parse():
    pts = int(input())
    return [
        vec(*map(fr.Fraction, input().strip().split()))
        for _ in range(pts)
    ]


if __name__ == "__main__":
    s = solve(parse())
    if s is None:
        print("unstable")
    else:
        print("{} .. {}".format(*s))

File: crane/crane.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <limits>
#include <iomanip>

static const double prec{100000};

using namespace std;

struct pt {
  double x, y;
  pt() : x{0}, y{0} {}
  pt(double x, double y) : x{x}, y{y} {}
  pt operator-() { return pt{this->x, this->y}; }
  void operator+=(pt b) { this->x += b.x; this->y += b.y; }
  void operator-=(pt b) { this->x -= b.x; this->y -= b.y; }
  void operator*=(double s) { this->x *= s; this->y *= s; }
  void operator/=(double s) { this->x /= s; this->y /= s; }
};

pt operator+(pt a, pt b) { return pt{a.x+b.x, a.y+b.y}; }
pt operator-(pt a, pt b) { return pt{a.x-b.x, a.y-b.y}; }
pt operator*(double s, pt a) { return pt{s*a.x, s*a.y}; }
pt operator*(pt a, double s) { return s * a; }
pt operator/(pt a, double s) { return pt{a.x/s, a.y/s}; }
double cross(pt a, pt b) { return a.x*b.y - b.x*a.y; }
pt moment(pt a, pt b) { return cross(a, b)/2 * (a+b)/3; }

struct interval {
  double low, high;
  interval(double low, double high) : low{low}, high{high} {}
};

interval intersect(interval a, interval b) {
  return interval{max(a.low, b.low), min(a.high, b.high)};
}

double area(vector<pt> pts) {
  double a{};
  for (size_t i{0}; i <= pts.size(); ++i)
    a += cross(pts[i], pts[(i+1)%pts.size()])/2;
  return a;
}

pt centroid(vector<pt> pts) {
  pt m{};
  for (size_t i{}; i <= pts.size(); ++i)
    m += moment(pts[i], pts[(i+1)%pts.size()]);
  return m/area(pts);
}

interval solve_inequality(double a, double b) {
  return a == 0
    ? (0 <= b
       ? interval{0, numeric_limits<double>::infinity()}
       : interval{0, -1})
    : (a > 0
       ? interval{0, b/a}
       : interval{max(0., b/a), numeric_limits<double>::infinity()}
       );

}

interval solve(vector<pt> pts) {
  double x{pts[0].x};
  double l{numeric_limits<double>::infinity()};
  double r{-numeric_limits<double>::infinity()};
  double m{abs(area(pts))};
  double c{centroid(pts).x};

  for (pt p : pts) {
    if (p.y) continue;
    if (p.x < l) l = p.x;
    if (p.x > r) r = p.x;
  }

  return intersect(solve_inequality(x-r, m*(r-c)),
                   solve_inequality(l-x, m*(c-l)));
}

int main() {
  vector<pt> pts;
  size_t count;
  cin >> count;
  for (size_t i{}; i < count; ++i) {
    pt p;
    cin >> p.x >> p.y;
    pts.push_back(p);
  }

  interval i{solve(pts)};
  if (i.low > i.high) {
    cout << "unstable" << endl;
  } else {
    cout
      << setprecision(0) << fixed
      << floor(round(i.low*prec)/prec) << " .. "
      << ceil(round(i.high*prec)/prec) << endl;
  }
}
